File: public/javascripts/model/queen.js
define([], function() {
    function Queen(colour, location, board) {
    
        function computeLinearMoves(getNext) {
            var testLocation = update(location);
            while(board.isValidLocation(testLocation)) {
                if (board.isOccupied(testLocation)) {
                    moves.forCapturing.push(testLocation);
                    break;
                } else {
                    moves.forPassing.push(testLocation);
                    testLocation = getNext(testLocation);
                }
            }
        }


        function computeMoves() {
            computeLinearMoves(upperRightDiagonal);
            computeLinearMoves(upperLeftDiagonal);
            computeLinearMoves(lowerRightDiagonal);
            computeLinearMoves(lowerLeftDiagonal);
            computeLinearMoves(rightRank);
            computeLinearMoves(leftRank);
            computeLinearMoves(upperFile);
            computeLinearMoves(lowerFile);
        }
        function initializeMoves() {
            moves.forCapturing = [];
            moves.forPassing = [];
        }

        this.getMoves = function() {
            initializeMoves();
            computeMoves();
            return moves;
        }
    }

    return Queen;
});

File: app/model/Queen.java
package com.chess.app;
import java.util.List;

public class Queen extends PieceImpl {
    public Queen(Colour colour, Location location) {
        super(colour, location); 
        directions = new DirectionsBuilder(colour)
            .with(new UpperRightDiagonal(Integer.MAX_VALUE))
            .with(new UpperLeftDiagonal(Integer.MAX_VALUE))
            .with(new LowerRightDiagonal(Integer.MAX_VALUE))
            .with(new LowerLeftDiagonal(Integer.MAX_VALUE))
            .with(new UpperFile(Integer.MAX_VALUE))
            .with(new LowerFile(Integer.MAX_VALUE))
            .with(new RightRank(Integer.MAX_VALUE))
            .with(new LeftRank(Integer.MAX_VALUE))
            .build();
    }
}
